kiosk/sources/extensions/owned_kiosk.move

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Description:
+///
+module kiosk::owned_kiosk {
+    use std::option::{Self, Option};
+    use sui::transfer;
+    use sui::kiosk::{Self, Kiosk, KioskOwnerCap};
+    use sui::object::{Self, ID, UID};
+    use sui::tx_context::{sender, TxContext};
+    use sui::dynamic_field as df;
+
+    const EIncorrectCapObject: u64 = 0;
+    const EIncorrectOwnedObject: u64 = 1;
+
+    /// A key-only wrapper for the KioskOwnerCap. Makes sure that the Kiosk can
+    /// not be traded altogether with its contents.
+    struct OwnedKiosk has key {
+        id: UID,
+        cap: Option<KioskOwnerCap>
+    }
+
+    /// The hot potato making sure the KioskOwnerCap is returned after borrowing.
+    struct Borrow { cap_id: ID, owned_id: ID }
+
+    /// The dynamic field to mark the Kiosk as owned (to allow guaranteed owner
+    /// checks through the Kiosk).
+    struct OwnerMarker has copy, store, drop {}
+
+    /// Wrap the KioskOwnerCap making the Kiosk "owned" and non-transferable.
+    public fun new(kiosk: &mut Kiosk, cap: KioskOwnerCap, ctx: &mut TxContext) {
+        let owner = sender(ctx);
+
+        // set the owner property of the Kiosk
+        kiosk::set_owner(kiosk, &cap, ctx);
+
+        // add the owner marker to the Kiosk; uses `_as_owner` to always pass,
+        // even if Kiosk "allow_extensions" is set to false
+        df::add(
+            kiosk::uid_mut_as_owner(kiosk, &cap),
+            OwnerMarker {},
+            owner
+        );
+
+        // wrap the Cap in the OwnedKiosk
+        transfer::transfer(OwnedKiosk {
+            id: object::new(ctx),
+            cap: option::some(cap)
+        }, sender(ctx));
+    }
+
+    /// Borrow the KioskOwnerCap from the OwnedKiosk object; Borrow hot-potato
+    /// makes sure that the Cap is returned via `return_cap` call.
+    public fun borrow_cap(self: &mut OwnedKiosk): (KioskOwnerCap, Borrow) {
+        let cap = option::extract(&mut self.cap);
+        let id = object::id(&cap);
+
+        (cap, Borrow {
+            owned_id: object::id(self),
+            cap_id: id
+        })
+    }
+
```

Comment:
I think we should also add a `borrow_cap(&OwnedKiosk): &KioskOwnerCap` for more convenient usage directly from Move. Perhaps the function above should be called `borrow_cap_mut`?

Reply:
Perhaps. We could give the same methods we have in the base Kiosk: `borrow`, `borrow_mut` and `borrow_val`.

Reply:
Would you mind adding this?